src/python/gudhi/barycenter.py

```diff
@@ -0,0 +1,233 @@
+
+    a = np.full(n+1, 1. / (n + m) )  
+    a[-1] = a[-1] * m                
+    b = np.full(m+1, 1. / (n + m) )  
+    b[-1] = b[-1] * n                
+    P = ot.emd(a=a, b=b, M=M)*(n+m)
+    # Note : it seems POT returns a permutation matrix in this situation, 
+    # ie a vertex of the constraint set (generically true).
+    if withcost:
+        cost = np.sum(np.multiply(P, M))
+    P[P < 0.5] = 0  # dirty trick to avoid some numerical issues... to improve.
+    res = np.nonzero(P)
+
+    # return the list of (i,j) such that P[i,j] > 0, 
+    #i.e. x_i is matched to y_j (should it be the diag). 
+    if withcost:
+        return np.column_stack(res), cost
+    
+    return np.column_stack(res)
+
+
+def lagrangian_barycenter(pdiagset, init=None, verbose=False):
+    """
+        Compute the estimated barycenter computed with the algorithm provided
+        by Turner et al (2014).
+        It is a local minimum of the corresponding Frechet function.
+    :param pdiagset: a list of size m containing numpy.array of shape (n x 2) 
+                        (n can variate), encoding a set of 
+                        persistence diagrams with only finite coordinates. 
+    :param init: The initial value for barycenter estimate. 
+                    If None, init is made on a random diagram from the dataset. 
+                    Otherwise, it must be an int 
+                    (then we init with diagset[init])
+                    or a (n x 2) numpy.array enconding 
+                    a persistence diagram with n points.
+    :param verbose: if True, returns additional information about the
+                        barycenter.
+    :returns: If not verbose (default), a numpy.array encoding
+                    the barycenter estimate 
+                    (local minima of the energy function). 
+                    If verbose, returns a couple (Y, log)
+                    where Y is the barycenter estimate,
+                    and log is a dict that contains additional informations:
+                    - groupings, a list of list of pairs (i,j),
+                        That is, G[k] = [(i, j) ...], where (i,j) indicates 
+                        that X[i] is matched to Y[j]
+                        if i > len(X) or j > len(Y), it means they 
+                        represent the diagonal.
+                    - energy, a float representing the Frechet 
+                                energy value obtained,
+                                that is the mean of squared distances 
+                                of observations to the output.
+                    - nb_iter, integer representing the number of iterations 
+                               performed before convergence of the algorithm.
+    """
+    X = pdiagset  # to shorten notations, not a copy
+    m = len(X)  # number of diagrams we are averaging
+    if m == 0:
+        print("Warning: computing barycenter of empty diag set. Returns None")
+        return None
```

Comment:
I haven't had time to think, are there obvious advantages returning None rather than an empty diagram for instance?

Reply:
Well, for me, the empty diagrams would be the barycenter of several empty diagrams, but averaging "no" diagrams might be a bit weird, but I agree that the empty diagram could be a correct answer (just like in R, when you say a sum over the empty set is 0). 
I just think that averaging an empty set of diagrams is a weird idea, so it would be adequate to have a warning there. If you're running Lloyd algo for K-Means for instance, and one centroid got "unaffected", you may not want to tell at the next iteration that it's "0" (or the emptydgm), right?